### PR TITLE
pipewire cleanup buildformat

### PR DESCRIPTION
**Pull request by columbarius**
This MR creates a seperate function to build pipewire formats.

### CODE REVIEW

File: src/screencast/pipewire_screencast.c [hunk 1]
```diff
@@ -26,6 +26,38 @@ static void writeFrameData(void *pwFramePointer, void *wlrFramePointer,
 	return;
 }
 
+static inline struct spa_pod *build_format(struct spa_pod_builder *b, uint32_t format, uint32_t format_without_alpha,
```
Comment (emersion) [inline src/screencast/pipewire_screencast.c]: Don't use `inline`, the compiler will automatically inline functions if it's worth it.

File: src/screencast/pipewire_screencast.c [hunk 2]
```diff
@@ -26,6 +26,38 @@ static void writeFrameData(void *pwFramePointer, void *wlrFramePointer,
...
+		uint32_t width, uint32_t height, uint32_t framerate)
+{
```
Comment (emersion) [inline src/screencast/pipewire_screencast.c:31]: Style: `{` on the prev line

File: src/screencast/pipewire_screencast.c [hunk 3]
```diff
@@ -26,19 +26,28 @@ static void writeFrameData(void *pwFramePointer, void *wlrFramePointer,
...
+	enum spa_video_format format_without_alpha = xdpw_format_pw_strip_alpha(format);
+
+	spa_pod_builder_push_object(b, &f[0], SPA_TYPE_OBJECT_Format, SPA_PARAM_EnumFormat);
 	spa_pod_builder_add(b, SPA_FORMAT_mediaType, SPA_POD_Id(SPA_MEDIA_TYPE_video), 0);
 	spa_pod_builder_add(b, SPA_FORMAT_mediaSubtype, SPA_POD_Id(SPA_MEDIA_SUBTYPE_raw), 0);
-	if (format_without_alpha != SPA_VIDEO_FORMAT_UNKNOWN) {
-		spa_pod_builder_add(b, SPA_FORMAT_VIDEO_format,
-			SPA_POD_CHOICE_ENUM_Id(3, format, format, format_without_alpha), 0);
+	/* format */
+	spa_pod_builder_prop(b, SPA_FORMAT_VIDEO_format, 0);
```
Comment (emersion) [inline src/screencast/pipewire_screencast.c:40]: IMHO this logic is less readable than the original one.

**Comment by emersion** [comment]
Needs a rebase

File: src/screencast/pipewire_screencast.c [hunk 4]
```diff
@@ -26,6 +26,44 @@ static void writeFrameData(void *pwFramePointer, void *wlrFramePointer,
 	return;
 }
 
+static inline struct spa_pod *build_format(struct spa_pod_builder *b, uint32_t format,
```
Comment (emersion) [inline src/screencast/pipewire_screencast.c:29]: Can't we use `enum spa_video_format` for the format here? That would make function calls slightly less error-prone.

File: src/screencast/pipewire_screencast.c (hunk 3 above)
Comment (columbarius) [reply src/screencast/pipewire_screencast.c:40]: I will do a similar logic for modifiers so i wanted to make things similar, but we can just use the old logic.

File: src/screencast/pipewire_screencast.c (hunk 4 above)
Comment (columbarius) [reply src/screencast/pipewire_screencast.c:29]: sure

**Comment by columbarius** [comment]
Thanks for the review!

File: src/screencast/pipewire_screencast.c (hunk 1 above)
Comment (columbarius) [reply src/screencast/pipewire_screencast.c]: done

File: src/screencast/pipewire_screencast.c (hunk 2 above)
Comment (columbarius) [reply src/screencast/pipewire_screencast.c]: done

File: src/screencast/pipewire_screencast.c (hunk 3 above)
Comment (columbarius) [reply src/screencast/pipewire_screencast.c]: added a commit which simplifies the path with alpha. we can squash them, but i didn't want to loose the old version for now.

File: src/screencast/pipewire_screencast.c (hunk 4 above)
Comment (columbarius) [reply src/screencast/pipewire_screencast.c]: done

**Comment by columbarius** [comment]
@emersion 
Rebase done

File: src/screencast/pipewire_screencast.c [hunk 5]
```diff
@@ -26,6 +26,38 @@ static void writeFrameData(void *pwFramePointer, void *wlrFramePointer,
...
+
+	enum spa_video_format format_without_alpha = xdpw_format_pw_strip_alpha(format);
+
+	spa_pod_builder_push_object(b, &f[0], SPA_TYPE_OBJECT_Format, SPA_PARAM_EnumFormat);
+	spa_pod_builder_add(b, SPA_FORMAT_mediaType, SPA_POD_Id(SPA_MEDIA_TYPE_video), 0);
+	spa_pod_builder_add(b, SPA_FORMAT_mediaSubtype, SPA_POD_Id(SPA_MEDIA_SUBTYPE_raw), 0);
+	/* format */
+	if (format_without_alpha == SPA_VIDEO_FORMAT_UNKNOWN) {
+		spa_pod_builder_prop(b, SPA_FORMAT_VIDEO_format, 0);
+		spa_pod_builder_id(b, format);
```
Comment (emersion) [inline src/screencast/pipewire_screencast.c:41]: Can this be simplified to the following?

```c
spa_pod_builder_add(b, SPA_FORMAT_VIDEO_format, SPA_POD_Id(format), 0);
```

Comment (columbarius) [reply src/screencast/pipewire_screencast.c]: yes

**Review by emersion — APPROVED** [review]
Thanks!

**Comment by columbarius** [comment]
Thanks for merging!

**Pull request merged**